### PR TITLE
Implement duck shaping on SymInts

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85808

Duck shaping says that when two input tensors have the same
size, we assume they are symbolically related.  This follows
the same optimization done by inductor.

This optimization is not done completely because we don't
currently install guards corresponding to the duck shape
relationships we created, but overall the guard propagation
for dynamic shape tracing is incomplete at the moment.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -274,17 +277,27 @@ def _get_key(self):
...
     def create_symint(self, name, val):
+        assert val >= 0
         if not HAS_SYMPY:
             raise RuntimeError("Need sympy installed to create symbolic shapes")
 
-        # Currently we don't put 0/1 specialization in guards but perhaps we should
+        # TODO: Put 0/1 specialization in guards
         if val == 0 or val == 1:
             return val
+        # This implements duck-shaping: input sizes that match are assigned
```
Comment (albanD) [inline torch/fx/experimental/symbolic_shapes.py:290]: This is a tradeoff to speed up tracing as we have less symint to handle at the cost of some potential recompilation?

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:290]: Yep

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:290]: It's also not just tracing speed; it's easier to generate kernels with more constraints like this

**Review by albanD — APPROVED** [review]
Sounds good

**Pull request closed without merging**